Web Apps: Populate value for unsupported question types

orangejenny opened the pull request:
http://manage.dimagi.com/default.asp?249088

Give unsupported question types a default response so they don't block form submission in web apps / app preview if they're required.

@dimagi/product / @ctsims Does this approach seem reasonable? The other suggestion on the ticket (supply a text input) would be more flexible in terms of passing validation conditions and making reasonable fake output, but it's also potentially a good deal more confusing. I'd also appreciate feedback on whether or not to include the "This question will appear..." message or if it's just clutter.

App Preview:
<img width="297" alt="screen shot 2018-08-09 at 2 09 31 pm" src="https://user-images.githubusercontent.com/1486591/43917697-02c0da98-9bdf-11e8-82b1-f7acd12dcfa7.png">

Form Data:
<img width="1118" alt="screen shot 2018-08-09 at 2 09 55 pm" src="https://user-images.githubusercontent.com/1486591/43917723-0d853f14-9bdf-11e8-838f-17d1fa03dc8f.png">

code buddies @pr33thi @snopoke 
web apps @jmtroth0

orangejenny committed:
Web Apps: Populate value for unsupported question types

djmore9 commented:
Thanks, Jenny.

I would get rid of the 'This question will appear...' message, and change
the Response to Not supported by web entry.

Dev




On Thu, Aug 9, 2018 at 2:25 PM, Jenny Schweers <notifications@github.com>
wrote:

> http://manage.dimagi.com/default.asp?249088
>
> Give unsupported question types a default response so they don't block
> form submission in web apps / app preview if they're required.
>
> @dimagi/product <https://github.com/orgs/dimagi/teams/product> / @ctsims
> <https://github.com/ctsims> Does this approach seem reasonable? The other
> suggestion on the ticket (supply a text input) would be more flexible in
> terms of passing validation conditions and making reasonable fake output,
> but it's also potentially a good deal more confusing. I'd also appreciate
> feedback on whether or not to include the "This question will appear..."
> message or if it's just clutter.
>
> [image: screen shot 2018-08-09 at 2 09 31 pm]
> <https://user-images.githubusercontent.com/1486591/43917697-02c0da98-9bdf-11e8-82b1-f7acd12dcfa7.png>
>
> [image: screen shot 2018-08-09 at 2 09 55 pm]
> <https://user-images.githubusercontent.com/1486591/43917723-0d853f14-9bdf-11e8-838f-17d1fa03dc8f.png>
>
> code buddies @pr33thi <https://github.com/pr33thi> @snopoke
> <https://github.com/snopoke>
> web apps @jmtroth0 <https://github.com/jmtroth0>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/dimagi/commcare-hq/pull/21498
> Commit Summary
>
>    - Web Apps: Populate value for unsupported question types
>
> File Changes
>
>    - *M* corehq/apps/cloudcare/static/cloudcare/js/form_entry/
>    entrycontrols_full.js
>    <https://github.com/dimagi/commcare-hq/pull/21498/files#diff-0> (1)
>    - *M* corehq/apps/cloudcare/templates/form_entry/templates.html
>    <https://github.com/dimagi/commcare-hq/pull/21498/files#diff-1> (13)
>
> Patch Links:
>
>    - https://github.com/dimagi/commcare-hq/pull/21498.patch
>    - https://github.com/dimagi/commcare-hq/pull/21498.diff
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/21498>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdarmbJ50jadX5fgGJIK0wZIP79wLF0xks5uPH6LgaJpZM4V2L67>
> .
>



-- 
Devendra More
Director of Product, SaaS
Dimagi, Inc | 585 Massachusetts Ave | Suite 3 | Cambridge, MA 02139
t: 617-649-2214 <(617)%20649-2214> | m: 857-415-9419 <(617)%20649-2214> |
Skype: dmore@dimagi.com
web: http://www.dimagi.com/

orangejenny committed:
Update copy

orangejenny commented:
@djmore9 updated

ctsims commented:
Hm, so I think that the failure modes associated with populating default
values with random data are likely enough that *some* form of user
intervention should be required.

Three options jump to mind for me
1) Do something arbitrary inline and tell the user what you did very
deliberately (this approach)
2) Do something arbitrary, but tell the user what it is in external docs
(IE: have a "mouseover"  ( ? ) circle that you can click for more details
that pops up a full page).
3) Require user input for the user to proceed<- I'd probably do this in the
form builder rather than in webapps, but we could make people give us a
"default value for unsupported platforms" value or something.

I'd prefer #2 here over not doing *anything* to tell the user, and might
even suggest that the form submission on HQ later could tell you that the
form was submitted in a weird way.


On Thu, Aug 9, 2018 at 3:47 PM, djmore9 <notifications@github.com> wrote:

> Thanks, Jenny.
>
> I would get rid of the 'This question will appear...' message, and change
> the Response to Not supported by web entry.
>
> Dev
>
>
>
>
> On Thu, Aug 9, 2018 at 2:25 PM, Jenny Schweers <notifications@github.com>
> wrote:
>
> > http://manage.dimagi.com/default.asp?249088
> >
> > Give unsupported question types a default response so they don't block
> > form submission in web apps / app preview if they're required.
> >
> > @dimagi/product <https://github.com/orgs/dimagi/teams/product> / @ctsims
> > <https://github.com/ctsims> Does this approach seem reasonable? The
> other
> > suggestion on the ticket (supply a text input) would be more flexible in
> > terms of passing validation conditions and making reasonable fake output,
> > but it's also potentially a good deal more confusing. I'd also appreciate
> > feedback on whether or not to include the "This question will appear..."
> > message or if it's just clutter.
> >
> > [image: screen shot 2018-08-09 at 2 09 31 pm]
> > <https://user-images.githubusercontent.com/1486591/
> 43917697-02c0da98-9bdf-11e8-82b1-f7acd12dcfa7.png>
> >
> > [image: screen shot 2018-08-09 at 2 09 55 pm]
> > <https://user-images.githubusercontent.com/1486591/
> 43917723-0d853f14-9bdf-11e8-838f-17d1fa03dc8f.png>
> >
> > code buddies @pr33thi <https://github.com/pr33thi> @snopoke
> > <https://github.com/snopoke>
> > web apps @jmtroth0 <https://github.com/jmtroth0>
> > ------------------------------
> > You can view, comment on, or merge this pull request online at:
> >
> > https://github.com/dimagi/commcare-hq/pull/21498
> > Commit Summary
> >
> > - Web Apps: Populate value for unsupported question types
> >
> > File Changes
> >
> > - *M* corehq/apps/cloudcare/static/cloudcare/js/form_entry/
> > entrycontrols_full.js
> > <https://github.com/dimagi/commcare-hq/pull/21498/files#diff-0> (1)
> > - *M* corehq/apps/cloudcare/templates/form_entry/templates.html
> > <https://github.com/dimagi/commcare-hq/pull/21498/files#diff-1> (13)
> >
> > Patch Links:
> >
> > - https://github.com/dimagi/commcare-hq/pull/21498.patch
> > - https://github.com/dimagi/commcare-hq/pull/21498.diff
> >
> > —
> > You are receiving this because you are on a team that was mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/dimagi/commcare-hq/pull/21498>, or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AdarmbJ50jadX5fgGJIK0wZIP79wLF0xks5uPH6LgaJpZM4V2L67>
> > .
> >
>
>
>
> --
> Devendra More
> Director of Product, SaaS
> Dimagi, Inc | 585 Massachusetts Ave | Suite 3 | Cambridge, MA 02139
> <https://maps.google.com/?q=585+Massachusetts+Ave+%7C+Suite+3+%7C+Cambridge,+MA+02139&entry=gmail&source=g>
> t: 617-649-2214 <(617)%20649-2214> | m: 857-415-9419 <(617)%20649-2214> |
> Skype: dmore@dimagi.com
> web: http://www.dimagi.com/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/21498#issuecomment-411875181>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJdusbJZ6xUKcqhUSXtlTxMUX6-dP5Gks5uPJHWgaJpZM4V2L67>
> .
>

orangejenny commented:
@ctsims I like the mouseover, it'd be less obtrusive. Regarding the form being submitted in a weird way, is it not already possible to tell that a form was submitted via web apps? Or do you mean some other kind of indicator that the submission is unusual, in this case because it was via web apps and contained unsupported question types?

ctsims commented:
Yeah, I mean specifically flagging that the submission may contain invalid
/ inconsistent data due to an answer to a web apps question that was
unsupported. Otherwise we could miss why a form / app was in a weirdly
impossible state.


On Thu, Aug 9, 2018 at 4:05 PM, Jenny Schweers <notifications@github.com>
wrote:

> @ctsims <https://github.com/ctsims> I like the mouseover, it'd be less
> obtrusive. Regarding the form being submitted in a weird way, is it not
> already possible to tell that a form was submitted via web apps? Or do you
> mean some other kind of indicator that the submission is unusual, in this
> case because it was via web apps and contained unsupported question types?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/21498#issuecomment-411880380>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJduu6GFGQa3hGsq6JsVZEmbHUvG_sYks5uPJYfgaJpZM4V2L67>
> .
>

ohmegasquared commented:
Also worthwhile to call out, in my test app I had created required questions that were not supported by web apps and there was no way to circumnavigate testing it without changing the app. (I was running into the issue for both date and time stamp and upload media.) 

From a UX perspective a good error message helps resolve the error at hand: "Sorry this isn't supported. Try XXX." Without any cues, the user is stuck in the same frustrating situation.

orangejenny committed:
Switch to modal

orangejenny commented:
How about this as a compromise? I added an info icon and made it pop up a modal, which is consistent with how we we display hint text. That gives me more space and is a click away, so I changed the text to an input that defaults to the same message.

The "default value for unsupported platforms" would be more robust for projects that use web apps as a standard workflow, but it's a bit esoteric for someone just trying to get through app preview. It would be a reasonable followup for this, it'd just change what the default value is.

<img width="295" alt="screen shot 2018-08-15 at 9 20 12 am" src="https://user-images.githubusercontent.com/1486591/44150562-d4477878-a06d-11e8-8fb9-23a1f557de8f.png">
<img width="294" alt="screen shot 2018-08-15 at 9 20 32 am" src="https://user-images.githubusercontent.com/1486591/44150563-d4650474-a06d-11e8-9706-b3daeddb1051.png">

ctsims commented:
This seems like an appropriate compromise. Thanks for updating this!

-Clayton

On Wed, Aug 15, 2018 at 9:38 AM, Jenny Schweers <notifications@github.com>
wrote:

> How about this as a compromise? I added an info icon and made it pop up a
> modal, which is consistent with how we we display hint text. That gives me
> more space and is a click away, so I changed the text to an input that
> defaults to the same message.
>
> The "default value for unsupported platforms" would be more robust for
> projects that use web apps as a standard workflow, but it's a bit esoteric
> for someone just trying to get through app preview. It would be a
> reasonable followup for this, it'd just change what the default value is.
>
> [image: screen shot 2018-08-15 at 9 20 12 am]
> <https://user-images.githubusercontent.com/1486591/44150562-d4477878-a06d-11e8-8fb9-23a1f557de8f.png>
>
> [image: screen shot 2018-08-15 at 9 20 32 am]
> <https://user-images.githubusercontent.com/1486591/44150563-d4650474-a06d-11e8-9706-b3daeddb1051.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/21498#issuecomment-413199931>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJdujbXJXz8xgUD53tTBqgN4TLHP7R2ks5uRCRGgaJpZM4V2L67>
> .
>

djmore9 commented:
Looks good. Are you concerned that users might know that info icon is
clickable? [This is a larger platform wide problem]

On Wed, Aug 15, 2018 at 9:38 AM, Jenny Schweers <notifications@github.com>
wrote:

> How about this as a compromise? I added an info icon and made it pop up a
> modal, which is consistent with how we we display hint text. That gives me
> more space and is a click away, so I changed the text to an input that
> defaults to the same message.
>
> The "default value for unsupported platforms" would be more robust for
> projects that use web apps as a standard workflow, but it's a bit esoteric
> for someone just trying to get through app preview. It would be a
> reasonable followup for this, it'd just change what the default value is.
>
> [image: screen shot 2018-08-15 at 9 20 12 am]
> <https://user-images.githubusercontent.com/1486591/44150562-d4477878-a06d-11e8-8fb9-23a1f557de8f.png>
>
> [image: screen shot 2018-08-15 at 9 20 32 am]
> <https://user-images.githubusercontent.com/1486591/44150563-d4650474-a06d-11e8-9706-b3daeddb1051.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/21498#issuecomment-413199931>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdarmZ_QH1qe6hQwH5aJEcCLCCKelx3aks5uRCRGgaJpZM4V2L67>
> .
>



-- 
Devendra More
Director of Product, SaaS
Dimagi, Inc | 585 Massachusetts Ave | Suite 3 | Cambridge, MA 02139
t: 617-649-2214 <(617)%20649-2214> | m: 857-415-9419 <(617)%20649-2214> |
Skype: dmore@dimagi.com
web: http://www.dimagi.com/

orangejenny commented:
@djmore9 What do you think about tweaking it like this? I'm not familiar with the larger problem, so not sure if this deals with it, but the (i) icon and its placement did look to me at first like a decoration or bullet point rather than something to click on.

<img width="235" alt="screen shot 2018-08-15 at 10 22 29 am" src="https://user-images.githubusercontent.com/1486591/44153095-40d39d62-a075-11e8-8751-aa61ed1130f5.png">

djmore9 commented:
Cool. I think, this works great. Thanks, Jenny.

Dev

On Wed, Aug 15, 2018 at 10:25 AM, Jenny Schweers <notifications@github.com>
wrote:

> @djmore9 <https://github.com/djmore9> What do you think about tweaking it
> like this? I'm not familiar with the larger problem, so not sure if this
> deals with it, but the (i) icon and its placement did look to me at first
> like a decoration or bullet point rather than something to click on.
>
> [image: screen shot 2018-08-15 at 10 22 29 am]
> <https://user-images.githubusercontent.com/1486591/44153095-40d39d62-a075-11e8-8751-aa61ed1130f5.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/21498#issuecomment-413214229>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AdarmSEuDOi1p11T8hK7nIJX3m45e7TDks5uRC9ugaJpZM4V2L67>
> .
>



-- 
Devendra More
Director of Product, SaaS
Dimagi, Inc | 585 Massachusetts Ave | Suite 3 | Cambridge, MA 02139
t: 617-649-2214 <(617)%20649-2214> | m: 857-415-9419 <(617)%20649-2214> |
Skype: dmore@dimagi.com
web: http://www.dimagi.com/

orangejenny committed:
Change icon symbol & placement

orangejenny commented:
ready for review @pr33thi @snopoke @jmtroth0
